feature: 355: snapshot with local file: run meshery without nats, output results into local file

n2h9 opened the pull request:
…

**Description**

This PR implements #355

Continuation on 
https://github.com/meshery/meshsync/pull/364 
and
https://github.com/meshery/meshsync/pull/361  

**Notes for Reviewers**

***Was able to achieve***
Run meshsync without nats and output to file as json (instead of nats);

***Import of result file into kanvas.***
~~When I try to upload result file to kanvas -> it doesn't work, as kanvas expects  file to be yml or yaml. Simple conversion in online editor from json to yaml didn't help, probably need to play a little with the format.~~

~~But If I take json of one resource (f.e. pod) andconvert it to yaml and import it to yaml -> it works.~~

~~So here need to output resource in the correct yaml list so that the output could be imported to kanvas directly.~~

Updated marshaling to yaml.

**[Signed commits](../CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.
 

<!--
Thank you for contributing to Meshery! 

Contributing Conventions:

1. Include descriptive PR titles with [<component-name>] prepended.
2. Build and test your changes before submitting a PR. 
3. Sign your commits

By following the community's contribution conventions upfront, the review process will 
be accelerated and your PR merged more quickly.
-->

n2h9 commented:
Hey hello, :wave: :innocent: ! 

Would appreciate if you @devhindo by chance  have some comments on this :blush:   thank you :bowing_woman:

lekaf974 commented on internal/file/file.go:
```diff
@@ -0,0 +1,35 @@
+	currentTime := time.Now()
+
+	// Format the current date in YYYYMMDD format
+	currentDate := currentTime.Format("20060102")
```
what if we want to have several snapshot a day ?

n2h9 replied:
Hey @lekaf974  hello :wave: :innocent: 

it will increment the name like 
- meshery-cluster-snapshot-20250330-11
- meshery-cluster-snapshot-20250330-12
...

![terminal 4](https://github.com/user-attachments/assets/19da55f9-ee79-477e-a927-dec3cea01906)


this is covered in the cycle in line 21.

lekaf974 replied:
Just see it is done in the loop below

lekaf974 commented on main.go:
```diff
@@ -98,24 +148,49 @@ func main() {
 
 	chPool := channels.NewChannelPool()
-	meshsyncHandler, err := meshsync.New(cfg, log, br, chPool)
+	meshsyncHandler, err := meshsync.New(cfg, log, br, fw, chPool)
```
Not a blocker but let's say we want to add another output mode (ex: s3, etc.) we will add another argument to the method ? 

For example using a map[string]interface{}  where we can add use OutputMode as key will allow to add other type without modify Method signature

n2h9 replied:
Hey @lekaf974 :wave: 

You are right! 

We are passing right now both broker (variable name _br_) and file writter (variable name _fw_). Which does not make sense as only one of them is in use.  And as you mentioned it is not extensible to further output methods. :cry: 

Agree it would be great to abstract output mode. F.e. in the way how you have suggested it with a map or with an interface :innocent: ! 

Thank you for the suggestion! :hugs:

lekaf974 replied:
So only one mode is in used at runtime ?

In such case looks a struct using strategy design pattern would be a good canditate here

lekaf974 commented:
Great Job

lekaf974 commented on internal/pipeline/handlers.go:
```diff
@@ -78,14 +79,44 @@ func (ri *RegisterInformer) publishItem(obj *unstructured.Unstructured, evtype b
+
+	}
+
+	if internalconfig.OutputMode == internalconfig.OutputModeNats {
+		err := ri.broker.Publish(config.PublishTo, &broker.Message{
+			ObjectType: broker.MeshSync,
+			EventType:  evtype,
+			Object:     k8sResource,
+		})
+		if err != nil {
+			ri.log.Error(ErrPublish(config.Name, err))
+			return err
+		}
+	}
+	if internalconfig.OutputMode == internalconfig.OutputModeFile {
+		_, err := ri.fileWriter.Write(k8sResource)
+		if err != nil {
+			ri.log.Error(ErrWriteFile(config.Name, err))
+			return err
+		}
```
if I am write here is where behavior change based on OutputMode this is where Strategy design pattern can make code cleaner

n2h9 replied:
yes, you are right :slightly_smiling_face:  this is the main place where the difference between output modes is handled :innocent: .

n2h9 replied:
Hey, @lekaf974 :innocent: hello :wave: ! 

I have implemented by your suggestion something similar to strategy pattern in this commit: https://github.com/meshery/meshsync/pull/386/commits/b887c6c044f26c766e9cf631d5aa6239906d82dd  :slightly_smiling_face: 
Will appreciate if you have any further suggestions on it :blush: 

Thank you :bowing_woman:

n2h9 replied:
Deal :smile: :handshake:  updated :white_check_mark: 

The only concerns is how it is used.
The previous version was used 
`output.Strategy`
and the current version is used
`output.OutputWriter`
so double "output" :woman_shrugging: 

--
So I went slightly further and made it just Writer.
so now it is used as `output.Writer`, please let me know if better to make it back OutputWriter :innocent:

lekaf974 replied:
Make sense

lekaf974 commented on docker-compose.yaml:
```diff
@@ -0,0 +1,15 @@
+# this docker compose is solely for local run with main_test.go
+services:
+  nats:
+    image: nats:2.8.2-alpine3.15
```
the image has 1 critical vunerabilites
![image](https://github.com/user-attachments/assets/dd037eaf-f232-4437-88b9-edf48043b991)
could we look for one with less vulnerability or none (2.8.2-linux) or move to nats 2.11 ?

n2h9 replied:
Hey @lekaf974 :wave: 

I used the version which is used in meshsery broker deployed when you apply meshery operator from https://github.com/meshery/meshery/tree/master/install 

<img src="https://github.com/user-attachments/assets/4fd475d9-27d6-4570-bd85-04fe3f1e6ebb" width="512"/>

But was not able to find by quick search where the version is defined :slightly_smiling_face: 

Sure we definitely can use the recent latest version :innocent: it totally makes sense. :+1: 
Updated to use 2.11 which is the latest according to https://hub.docker.com/_/nats 

By the way how do you check the image vulnerabilities, is it some separate tool, or it is available somewhere in docker hub? :innocent:

lekaf974 replied:
available in hob docker directly now https://hub.docker.com/_/nats/tags?name=2.11

marblom007 commented:
A few merge conflicts need to be addressed, but I think this PR is ready after that...

marblom007 commented:
I merged these conflicts, @n2h9. I think that I did it correctly, but I see that the integration test is failing... 🤔

n2h9 commented:
> I merged these conflicts, @n2h9. I think that I did it correctly, but I see that the integration test is failing... 🤔

right, let me double check on this please . . .

n2h9 commented:
> > I merged these conflicts, @n2h9. I think that I did it correctly, but I see that the integration test is failing... 🤔
> 
> right, let me double check on this please . . .

Hey @marblom007 hello :wave: :innocent: 
It was a missing import, updated, should be all right now :blush:  

Thank you :bowing_woman:

n2h9 commented:
> > > I merged these conflicts, @n2h9. I think that I did it correctly, but I see that the integration test is failing... 🤔
> > 
> > 
> > right, let me double check on this please . . .
> 
> Hey @marblom007 hello 👋 😇 It was a missing import, updated, should be all right now 😊
> 
> Thank you 🙇‍♀️
> 
> --
> 
> Ha 😄 looks like it is still failing 😢 Let me triple check . . .

Ok, this one is interesting:
```sh
=== RUN   TestWithNatsIntegration
time="2025-04-18T18:32:45Z" level=error msg="Missing or outdated CRD. " app=meshsync code=1000 probable-cause="Missing or outdated CRD." severity=2 short-description="Error while initializing MeshSync configuration. .Config File \"meshsync_config\" Not Found in \"[/home/runner/.meshery]\"" suggested-remediation="Confirm that meshsyncs custom resource is present in the cluster."
```

>.Config File \"meshsync_config\" Not Found in \"[/home/runner/.meshery]\

this is related to this comment in [internal/config/config.go:16](https://github.com/meshery/meshsync/pull/386/files#diff-54c7c1af5fa8d5db4dc49f0e8e80e93ba2b1183ba4d5c9e2e5729e6deae6a3cdR16)

I inherited this change from this pr: https://github.com/meshery/meshsync/pull/361 

Let me double check with Lee if we need this and come back you @marblom007  :hugs:

n2h9 commented:
> 
> > .Config File "meshsync_config" Not Found in "[/home/runner/.meshery]\
> 
> this is related to this comment in [internal/config/config.go:16](https://github.com/meshery/meshsync/pull/386/files#diff-54c7c1af5fa8d5db4dc49f0e8e80e93ba2b1183ba4d5c9e2e5729e6deae6a3cdR16)
> 
> I inherited this change from this pr: #361
> 
> Let me double check with Lee if we need this and come back you @marblom007 🤗

I will probably need to add a code here, which creates the folder `$HOME/.meshery`. I mean not before test, but in the meshsync itself

n2h9 commented:
> I will probably need to add a code here, which creates the folder $HOME/.meshery. I mean not before test, but in the meshsync itself.
And also there are fresh lint errors 🙂
Let me fix this two issues and come back to you with updates 😇

Hey hello :wave: :innocent: ! 

I have updated as mentioned above to:
- create $HOME/.meshery folder if it not exists;
- and fixed lint errors;

and also added test cases to ensure current nats integration is not broken :smile: 

---

I would like also: 
- to fix a cyclomatic complexity in main, as it is now huge after my changes (it was big but now huge :smile: )
- to double check the actual functionality of output to local file, as I was mostly concentrated last couple days on how to check that existing set up is not broken; 

so will keep it as draft for now for a little longer, will mark as ready soon :innocent: 

thank you :bowing_woman:

n2h9 commented:
> to fix a cyclomatic complexity in main, as it is now huge after my changes (it was big but now huge 😄 )

I started to do it, and updated couple paths which was added by me, but probably better not to do it in this pr. As it will require almost complete refactoring of main file. Just to avoid possible regression, not related to functionality. 
Let's fix cyclomatic complexity in a separate pr when this one will be merged :innocent: 

--
By the way the average complexity for the main package dropped below 7, so this one is done :white_check_mark:

n2h9 commented:
Added couple test cases which do checks when output mode is file;

leecalcote commented:
Some good progress here. 🥇

lekaf974 reviewed (commented):
Sorry for delay I forgot to submit the review

n2h9 commented on main.go:
```diff
@@ -74,7 +65,7 @@ func mainWithExitCode() int {
 	if err != nil {
-		fmt.Println(err)
+		log.Error(err)
 		return 1
```
Sure thing :innocent: 
updated :white_check_mark: 

thank you :bowing_woman:

lekaf974 commented on internal/output/processor.go:
```diff
@@ -0,0 +1,27 @@
+)
+
+type Processor struct {
+	strategy Writer
```
```suggestion
	output Writer
```

lekaf974 replied:
Suggestion to keep align with the rename of strategy to output.Writer.

n2h9 replied:
Sure thing :hugs: 
Updated :white_check_mark: 

Thank you :bowing_woman:

lekaf974 commented on internal/output/processor.go:
```diff
@@ -0,0 +1,27 @@
+	return &Processor{}
+}
+
+func (p *Processor) SetStrategy(strategy Writer) {
```
```suggestion
func (p *Processor) SetOutput(output Writer) {
```
same idea

n2h9 replied:
Updated :white_check_mark: 

Thank you :bowing_woman:

lekaf974 commented on meshsync/meshsync.go:
```diff
@@ -49,7 +51,7 @@ func GetListOptionsFunc(config config.Handler) (func(*v1.ListOptions), error) {
 }
 
-func New(config config.Handler, log logger.Handler, br broker.Handler, pool map[string]channels.GenericChannel) (*Handler, error) {
+func New(config config.Handler, log logger.Handler, br broker.Handler, os output.Writer, pool map[string]channels.GenericChannel) (*Handler, error) {
```
```suggestion
func New(config config.Handler, log logger.Handler, br broker.Handler, op output.Writer, pool map[string]channels.GenericChannel) (*Handler, error) {
```
nit

n2h9 replied:
Updated :white_check_mark: 

But instead of `op` (which probably stands for **o**utput **p**rocessor or maybe **o**ut**p**ut) I use `ow`  which stands for **o**utput **w**riter. And  in the same file I further renamed `outputProcessor` property (where this `ow` (former `os`) is assigned) to be `outputWriter`, it looks like makes more sense in the current context.  

Thank you :bowing_woman:

lekaf974 reviewed (approved):
Added small comments but LGTM

lekaf974 reviewed (approved):
LGTM

leecalcote commented:
Docs forthcoming separately.
